pipeline/relation_embedder/path_concatenator/src/test/scala/weco/pipeline/path_concatenator/PathConcatenatorWorkerServiceTest.scala

```diff
@@ -0,0 +1,160 @@
+  */
+class PathConcatenatorWorkerServiceTest
+    extends AnyFunSpec
+    with WorkGenerators
+    with PipelineStorageStreamFixtures
+    with Eventually
+    with IntegrationPatience
+    with Akka
+    with ElasticsearchFixtures
+    with IndexFixtures {
+
+  it(
+    "updates a work and its children, sending all their paths to the downstream queue") {
+    val works = List(
+      work("a/b"),
+      work("b/c"),
+      work("c/d"),
+    )
+
+    withWorkerService(works) {
+      case (QueuePair(queue, dlq), index, downstreamMessageSender) =>
+        sendNotificationToSQS(queue = queue, body = "b/c")
+        eventually {
+          assertQueueEmpty(queue)
+          assertQueueEmpty(dlq)
+          assertIndexContainsPaths(
+            index,
+            Map(
+              works(1).id -> "a/b/c",
+              works(2).id -> "a/b/c/d"
+            ))
+          assertQueueContainsPaths(
+            downstreamMessageSender,
+            List("b/c", "a/b/c", "a/b/c/d"))
+        }
+    }
+  }
+
+  it("sends the input path to the downstream queue when there is nothing to do") {
+    val works = List(
+      work("a/b"),
+      work("c/d"),
+    )
+
+    withWorkerService(works) {
+      case (QueuePair(queue, dlq), index, downstreamMessageSender) =>
+        sendNotificationToSQS(queue = queue, body = "c/d")
+        eventually {
+          assertQueueEmpty(queue)
+          assertQueueEmpty(dlq)
+          index shouldBe empty
+          assertQueueContainsPaths(downstreamMessageSender, List("c/d"))
+        }
+    }
+  }
+
+  it("sends the input path to the downstream queue if there are data errors") {
+    val works = List(
+      work("a/b"),
+      work("b/b"),
+      work("b/c"),
+    )
+
+    withWorkerService(works) {
+      case (QueuePair(queue, dlq), index, downstreamMessageSender) =>
+        sendNotificationToSQS(queue = queue, body = "b/c")
+        eventually {
+          assertQueueEmpty(queue)
+          assertQueueEmpty(dlq)
+          index shouldBe empty
+          assertQueueContainsPaths(downstreamMessageSender, List("b/c"))
+        }
+    }
+  }
+
+  private def work(path: String): Work.Visible[Merged] =
+    mergedWork(createSourceIdentifierWith(value = path))
+      .collectionPath(CollectionPath(path = path))
+      .title(path)
+
+  private def assertIndexContainsPaths(
+    inMemoryIndex: mutable.Map[String, Work[Merged]],
+    pathMap: Map[String, String]) =
+    inMemoryIndex map {
+      case (workId, work) =>
+        (workId, work.data.collectionPath.get.path)
+    } should contain theSameElementsAs pathMap
+
+  private def assertQueueContainsPaths(sender: MemoryMessageSender,
+                                       expectedPaths: List[String]) =
+    sender.messages.map(_.body) should contain theSameElementsAs expectedPaths
+
+  private def storeWorks(index: Index, works: List[Work[Merged]]): Assertion =
+    insertIntoElasticsearch(index, works: _*)
```

Comment:
I'm not a massive fan of this sort of very light helper, because it adds a layer of obfuscation – every test will have a slightly differently named method doing the same thing, and I find it harder to follow.

Not a must change, just a note.

Reply:
I agree.  This was actually copied from the Relation Embedder, in an attempt at fostering consistency.